snc-library.sh

```diff
@@ -253,3 +253,52 @@ function delete_operator() {
         # Wait until the operator pod is deleted before trying to delete the resources it manages
         ${OC} wait --for=delete pod/${pod} --timeout=120s -n ${namespace} || ${OC} delete pod/${pod} --grace-period=0 --force -n ${namespace} || true
 }
+
+function operators_available() {
+        # Check the cluster operator output, status for available should be true for all operators
+        ${OC} get co -ojsonpath='{.items[*].status.conditions[?(@.type=="Available")].status}' | grep -v False
+}
+
+function operators_progressing() {
+        # Check the cluster operator output, status for progressing should be false for all operators
+        ${OC} get co -ojsonpath='{.items[*].status.conditions[?(@.type=="Progressing")].status}' | grep -v True
```

Comment:
Sorry, won't have time for a proper review right now, but the logic seems off in these `operators_*()` functions. They check that that at least one operator status is not `False` or `True` depending on the function. I tried with the following test case, hopefully I did not make mistakes
```
function test_grep_false() {
	grep -v False <<EOF
1: False
2: False
3: True
EOF
}

function test_grep_true() {
	grep -v True <<EOF
1: False
2: False
3: True
EOF
}

if test_grep_false; then 
	echo "if (test_grep_false) succeeded"
else 
	echo "if (test_grep_false) failed"
fi

echo ""

if test_grep_true; then 
	# the test fails if all operators have a "True" status
	# the test succeeds if at least one operator is "False"
	echo "if (test_grep_true) succeeded"
else 
	echo "if (test_grep_true) failed"
fi
```

Reply:
Regarding naming of the functions, when the comment is `# Check the cluster operator output, status for progressing should be false for all operators ` , the function could be named "no_operators_progressing"?

Reply:
@cfergeau I agree about the naming and after experimenting lot of different stuff looks like following is provide us correct result. Here you can change any value and see if the cluster status. If you have better suggestion how to achieve in bash then I would like to explore that. (I am not very good in bash)

```bash
#!/bin/bash

set -exuo pipefail

export LC_ALL=C.UTF-8
export LANG=C.UTF-8

function all_available() {
        grep -v True >/dev/null <<EOF
True True True True
EOF
}

function none_progressing() {
        grep -v False >/dev/null <<EOF
False False False True
EOF
}

function none_degrading() {
        grep -v False >/dev/null <<EOF
True False False False
EOF
}

function check_cluster() {
        all_available
        a=$?
        none_progressing
        b=$?
        none_degrading
        c=$?
        echo $((a + b + c))
}

if [[ $(check_cluster) -ne 3 ]]; then
        echo "cluster unstable"
else
        echo "cluster stable"
fi
```

Reply:
Ah ok, you get a single line from `oc`, not multiple lines as in my test case. So my test case is indeed incorrect :)
For `none_progressing`/`none_degrading`, we want to check that the line we are processing does not contain any `True` occurrences. `grep True` will tell us if it does contain `True`, and `grep -v True` will tell us if there are no occurrences of `True`, so I think`none_progressing`/`none_degrading` should be doing this rather than `grep -v False`.

Similarly, for `all_available`, we want to make sure there are no `False` occurrences on the line, so `grep -v False`.

In short, your initial code seems correct apart from the naming, and my comments were misleading as I was not testing with the correct data

```bash
function all_operators_available() {
        # Check the cluster operator output, status for available should be true for all operators
        ${OC} get co -ojsonpath='{.items[*].status.conditions[?(@.type=="Available")].status}' | grep -v False
}

function no_operators_progressing() {
        # Check the cluster operator output, status for progressing should be false for all operators
        ${OC} get co -ojsonpath='{.items[*].status.conditions[?(@.type=="Progressing")].status}' | grep -v True
}

function no_operators_degraded() {
        # Check the cluster operator output, status for available should be false for all operators
        ${OC} get co -ojsonpath='{.items[*].status.conditions[?(@.type=="Degraded")].status}' | grep -v True
}
```